cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -42,13 +42,1440 @@
+    RAFT_EXPECTS(bytes_read == static_cast<ssize_t>(chunk_size),
+                 "Incomplete read from file. Expected %zu bytes, got %zd at offset %lu",
+                 chunk_size,
+                 bytes_read,
+                 file_pos);
+
+    bytes_remaining -= chunk_size;
+    offset += chunk_size;
+  }
+}
+
+// ACE: Write large file in chunks
+template <typename DataT, typename IdxT>
+void ace_write_large_file(const file_descriptor& fd,
+                          const void* data_ptr,
+                          const size_t total_bytes,
+                          const uint64_t file_offset)
+{
+  RAFT_EXPECTS(total_bytes > 0, "Total bytes must be greater than 0");
+  RAFT_EXPECTS(data_ptr != nullptr, "Data pointer must not be nullptr");
+  RAFT_EXPECTS(fd.is_valid(), "File descriptor must be valid");
+
+  const size_t write_chunk_size = 1024 * 1024 * 1024;
+  size_t bytes_remaining        = total_bytes;
+  size_t offset                 = 0;
+
+  while (bytes_remaining > 0) {
+    const size_t chunk_size = std::min(write_chunk_size, bytes_remaining);
+    const uint64_t file_pos = file_offset + offset;
+    const ssize_t chunk_written =
+      pwrite(fd.get(), reinterpret_cast<const char*>(data_ptr) + offset, chunk_size, file_pos);
+
+    RAFT_EXPECTS(
+      chunk_written != -1, "Failed to write to file at offset %lu: %s", file_pos, strerror(errno));
+    RAFT_EXPECTS(chunk_written == static_cast<ssize_t>(chunk_size),
+                 "Incomplete write to file. Expected %zu bytes, wrote %zd at offset %lu",
+                 chunk_size,
+                 chunk_written,
+                 file_pos);
+
+    bytes_remaining -= chunk_size;
+    offset += chunk_size;
+  }
+}
+
+// ACE: Get partition labels for partitioned approach
+template <typename DataT, typename IdxT, typename Accessor>
+void ace_get_partition_labels(
+  raft::resources const& res,
+  raft::mdspan<const DataT, raft::matrix_extent<int64_t>, raft::row_major, Accessor> dataset,
+  raft::host_matrix_view<IdxT, int64_t, raft::row_major> partition_labels,
+  raft::host_matrix_view<IdxT, int64_t, raft::row_major> partition_histogram,
+  uint64_t min_partition_size,
+  double sampling_rate = 0.01)
+{
+  uint64_t dataset_size = dataset.extent(0);
+  uint64_t dataset_dim  = dataset.extent(1);
+  uint64_t labels_size  = partition_labels.extent(0);
+  uint64_t labels_dim   = partition_labels.extent(1);
+  RAFT_EXPECTS(dataset_size == labels_size, "Dataset size must match partition labels extent");
+  uint64_t n_partitions = partition_histogram.extent(0);
+  RAFT_EXPECTS(labels_dim == 2, "Labels must have 2 columns");
+  RAFT_EXPECTS(partition_histogram.extent(1) == 2, "Partition histogram must have 2 columns");
+  cudaStream_t stream = raft::resource::get_cuda_stream(res);
+
+  // Sampling vectors from dataset
+  uint64_t n_samples = dataset_size * sampling_rate;
+  if (n_samples < 100 * n_partitions) { n_samples = 100 * n_partitions; }
```

Comment:
Here 100 is the min_samples_per_partition. I suggest to introduce a local constant to name it as such. Otherwise 100 is a good value, it is within the range of min/max points per cluster that faiss uses https://faiss.ai/cpp_api/struct/structfaiss_1_1ClusteringParameters.html


```suggestion
const uint64_t min_samples_per_partition = 100;
  if (n_samples < min_samples_per_partition * n_partitions) { n_samples = min_samples_per_partition * n_partitions; }
```